### PR TITLE
expose OpenTracing and API versioning capabilities

### PR DESCRIPTION
**Pull request by bensigelman**
Per https://github.com/opentracing/opentracing.github.io/issues/33 and https://github.com/opentracing/opentracing-go/pull/51#discussion_r52555910

@bcronin @yurishkuro

### CODE REVIEW

File: opentracing/harness/api_check.py [hunk 1]
```diff
@@ -46,7 +46,7 @@ def test_start_trace(self):
         with tracer.start_trace(operation_name='Fry',
                                 tags={'birthday': 'August 14 1974'}) as span:
             span.log_event('birthplace',
-		           payload={'hospital': 'Brooklyn Pre-Med Hospital',
+                   payload={'hospital': 'Brooklyn Pre-Med Hospital',
```
Comment (bensigelman) [inline opentracing/harness/api_check.py]: (I delinted and found some spaces/tabs things... probably my copy of vim)

File: opentracing/harness/api_check.py [hunk 2]
```diff
@@ -146,3 +146,8 @@ def test_binary_propagation(self):
                 trace_attributes=bin_attrs
             ) as reassembled_span:
                 reassembled_span.set_trace_attribute('middle-name', 'Rodriguez')
+
+    def test_implementation_id(self):
+        impl_id = self.tracer().implementation_id()
+        assert len(impl_id.name) > 0
```
Comment (bensigelman) [inline opentracing/harness/api_check.py]: not sure what else to check here...

File: opentracing/tracer.py [hunk 3]
```diff
@@ -82,3 +83,24 @@ def close(self):
...
+
+
+# Per collections.namedtuple, the below defines a class called ImplementationID
+# which can be initialized via positional parameters or a kwargs-style
+# initializer.
+#
+#     dapper_impl = ImplementationID('dapper', '0.1.2')
+#     zipkin_impl = ImplementationID(name='zipkin', version='0.3.4')
+#
+ImplementationID = namedtuple('ImplementationID', 'name version')
```
Comment (bensigelman) [inline opentracing/tracer.py]: I like `namedtuple` (makes me feel like I'm in a language that's not python :), but I could be talked into returning an ad hoc string map, too.

File: opentracing/tracer.py [hunk 4]
```diff
@@ -82,3 +83,24 @@ def close(self):
...
+
+# The SemVer (http://semver.org/) of the OpenTracing Python API.
+OPENTRACING_PYTHON_SEMVER = '0.9.0'
```
Comment (bensigelman) [inline opentracing/tracer.py]: @bcronin do we need anything fancier than this? I think the biggest challenge will be remembering to update it :-/

Of course we can automate that process to a certain extent...

Comment (yurishkuro) [inline opentracing/tracer.py]: I would put it at the top of the file.

> the biggest challenge will be remembering to update it

:-)  I remember seeing somewhere a lib was taking a hash of its public API and having a unit test comparing that hash with a constant, and a comment in upper case "if this hash changed the version needs to change". If I have any grey hair, that's when it started.

File: opentracing/tracer.py (hunk 3 above)
Comment (yurishkuro) [inline opentracing/tracer.py]: sgtm - it's also immutable. Not sure we need the comment in L96-98.

**Comment by yurishkuro** [comment]
+1

File: opentracing/tracer.py (hunk 4 above)
Comment (bcronin) [inline opentracing/tracer.py]: (Preface: I do not know python very well, so this might be just a language misunderstanding on my part...)

What I was imagining was given a `Tracer` object, there should be some way to query the supported OpenTracing API version of *that object*.  Having a global does not accomplish quite the same thing.  

At a high-level the problem is the case (impossible in some setups, but not all) were a single logical application has been bound, for one reason or another, to multiple versions of OpenTracing (via use of third-party libraries, nested dependencies, dynamically loaded code, etc.). This case is important because removing the multiple versions is not always something within an individual developer's control (e.g. nested dependencies and third-party code).

In Node.js for example, as far as I know it is fully possible to have two references to two different versions of the same package within a single logical application or service (there will be two distinct copies of the package within the loaded code; to help imagine the real-world scenario think of an application that relies on a framework that relies on third-party extensions that rely on third-party libraries -- and in that complex dependency tree someone has not updated their code to the latest version of OpenTracing but others have). The version check needs to be done on the actual `Tracer` object being used - and can't be a global property as the global property only would reflect the version of *the locally referenced package*, not necessarily *the version that created the Tracer object* in question. 

To clarify this is not Node.js specific, consider a C++ application that dynamically loads a library (i.e. a DLL/.so  plugin of some sort) that, as part of the plugin interface, takes a handle to the active Tracer. It may be interested in knowing the version of that runtime *Tracer object* it has been handed, not the version of OpenTracing the DLL itself was compiled against, to be sure the host application is compatible.  A global check would only check what it had been compiled against.

I certainly wouldn't argue these are going to be a common cases but to achieve the intent, I believe the version info has to be a property of the Tracer object, not a package/module/library-level property.

Comment (bensigelman) [inline opentracing/tracer.py]: @bcronin I see what you're saying. If we're going through the trouble of including this API in the first place, we may as well do it right. I'm not at my laptop right now, but I will think about it... It's annoying that Tracer impls need not actually inherit from Tracer (due to the usual dynamic typing situation)... If that weren't the case, this would be straightforward! Maybe we should just require it in a clearly worded comment or something...

**Comment by bensigelman** [comment]
@bcronin how about this latest version?

**Comment by bcronin** [comment]
The update addresses the "two copies" scenario I was concerned with, so now I'm descending to the next level of detail at this point (i.e. naming)...

I can't help but think it'd be a lot more straightforward to just have:

```python
def version():
    return '0.9.0'
```

* TL;DR: It's simpler :)
* As a blanket generalization, I'd argue making it a simple method translates to a more uniform interface across languages (e.g. I'm thinking of C++ and Go)
* In reality, I'd argue fairly strongly it is not a burden to implementors to have to implement this method! :) Worrying about an abstract base class just to avoid implementors having to implement a method returning a fixed string, IMHO, adds abstraction complexity and no actual, real-world value.  Also, I'd argue, OpenTracing API versions shouldn't be changing *that* often that keeping this value up to date should be any maintenance burden (note: this claim does *not* apply to the *implementation version* which in theory changes with every push by the implementator)
* `OPENTRACING_PYTHON_API_SEMVER` seems awkwardly verbose: the OPENTRACING and PYTHON part should be self-evident. I think that it's a *semantic* version is a documentation point (and I tend thinking using straightforward, full words is better than abbreviations when it's possible - i.e. `version`). 
* I also think including "API" in the name is unnecessary given the presence of something called `implementation_id` (this is harder to objectively argue, but I prefer the short name and would be surprised if consumers of OpenTracing found `version` confusing). That said, I believe I would still find a way to sleep soundly at night if it was called `api_version` :)

All and all, I think `Tracer.version()` would be simpler but I realize it's a large part just opinion and am not overly concerned about this.

**Comment by bensigelman** [comment]
Having slept on this, the only thing I really want to expose at this point is the semver of the per-platform OpenTracing API. I don't want the OT implementor to have to remember to update same, hence the use of the constant rather than a method which should be overridden.

Note that the per-platform semvers will not be in lock-step across the various platforms (if we refactor the API in one language, its semver will need to change, but there's no reason that the others must follow suit). Someday it would be nice to semver the platform-indepedent OpenTracing specification, but for now it's not formal enough to make a semver meaningful (in my opinion, anyway).

Thoughts?

In the meantime, I will trash the impl-id stuff entirely.

**Comment by bensigelman** [comment]
(trashing complete!)

**Comment by bcronin** [comment]
Ah, ok - the intent is to expose the _platform_ API version which may be independent of the spec version.  That still allows for "two copies" case to be handled, so I'm fine with this.

**Comment by yurishkuro** [comment]
Github confused me with the ordering of the comments...

 1. I prefer a method to a class constant
 1. I don't like that method to be returning a version string. It reminds me of that horrible user-agent parsing that JavaScript used to do to determine browser capabilities.

**Comment by bensigelman** [comment]
@yurishkuro 

1. The class constant was supposed to make it harder for an implementor to imagine that they're supposed to override in their subclass. Kind of like declaring a method `final` in Java (or something), but with static content. No super-strong feelings about that, though, and the lack of a docstring is annoying with the constant.
1. Are you suggesting a version tuple (of ints), or are you suggesting a struct-like thing with only a single version field for now? Or something else?

**Comment by yurishkuro** [comment]
@bensigelman 

 1. Not sure I understand the scenario. If tracer lib imports v.0.8 of opentracing API and extends the Tracer, it inherits the method with the correct version. Are we saying that the end user app has v.0.9 opentracing lib, but tracer impl written against 0.8? If that's the case, we can make the method abstract in the opentracing tracer, but that's not going to protect against implementor forgetting to bump the version. And I would also think the tracer's dependency should be `<0.9` in this case.

 2. I meant that exposing capabilities via version seems like a very fragile method. I can't imagine how ugly the client code will look doing those checks. The usual approach to combat API version changes is to forward or backport the API via a shim. E.g. if the instrumentation is done via v.0.9 but the bound tracer is only v.0.8 than the tracer should be wrapped in a decorator..

But I have to admit that I haven't had the pleasure of writing such spaghetti code in dynamic langauges and checking versions, so I don't want to `-1` this if people feel strongly that it's needed.

**Comment by bensigelman** [comment]
@yurishkuro 

1. I'm just saying that this particular version number is a property *of the API*, not of the implementation. So there's no need for the impl to override it, and thus no need to make it a method in an interface where all of the other methods are meant to be overridden.
1. A capabilities API would be more robust but also easier to regret / need to remove. I was trying to think of the smallest version-related change (i.e., something to address https://github.com/opentracing/opentracing.github.io/issues/33) that we would probably not want to remove in the future.

Thoughts? I am not attached to this PR but wanted to take a stab at https://github.com/opentracing/opentracing.github.io/issues/33 without adding a ton of complexity.

**Comment by yurishkuro** [comment]
1. It's a question of the use case. If an app manages to load conflicting versions of OT API lib and tracer implementation, and tracer simply refers to a constant in the API, then it's not representing what it actually implemented.  The scenario is hard to imagine in java/go, unusual in Python under normal `pip`-based dependency management, and I have no idea about JS.
  1. Fair enough. If (1) changes your mind to use a method, I'd call it `api_version`, not just `version`.

**Comment by bensigelman** [comment]
JS is a free-for-all... @bcronin actually understands it (I do not; or not well). I'll sleep on this whole thing... part of me wants to back it out until we have an actual specific problem to solve in the wild.

**Comment by bcronin** [comment]
The spirit of #33 was to provide a simple means of accessing the version of the OpenTracing standard being implemented against. If there also needs to be a separate notion of a platform version that can vary independently, frankly I think that adds additional complexity counter to the original goal and I suspect the intended value gets lost.

I think there's some (bear with a bit of soapboxing here to make myself feel better) danger in waiting until there's an actual specific problem to solve in the wild as that sort is missing a large part of adding versioning from the start. (Also, for the record most of forward/backwards compatibility experience here has to do with drivers and plug-ins in C++ SDK architectures, not JavaScript, but the API scope here is much smaller so I'm totally willing to concede that I may be overthinking it.)

Anyway, my soapboxing done and stepping back, there are things I'd rather prioritize here and, in practice, even if it's not elegant, people almost always find *some* way to work around version incompatibilities, so I think it'd be net-best to close #33.

**Comment by bensigelman** [comment]
(withdrawing per the fate of opentracing/opentracing.github.io#33)
